tuxemon/core/states/world/worldstate.py

```diff
@@ -618,13 +618,19 @@ def tile_blocks(self, entity, tile):
         :rtype: bool
         """
         if entity and tile and 'blocks' in tile:
-            for block in tile['blocks'].split():
-                if block == 'players' and entity.isplayer:
-                    return True
-                if block == 'npcs' and not entity.isplayer:
-                    return True
-                if block == entity.slug:
-                    return True
+            for block in tile['blocks']:
+                if block == 'players':
+                    if entity.isplayer:
+                        return True
+                elif block == 'npcs':
+                    if not entity.isplayer:
+                        return True
+                elif block in self.npcs:
```

Comment:
ok, this will check for slugs, which is great, but its still in "the hot path" for collision detection

Reply:
IIRC 'if x in y' is a cheap check in Python, so hopefully this doesn't add any extra load. If I checked the NPC database here, it would include all NPC's defined in the game not just those present on the map... the check would thus be more expensive as there are many extra keys to needlessly scan through.

It should be kept in mind that those checks are only preformed if a player or NPC is walking near a collision object that has the `block` parameter set: None of this is executed if the right conditions aren't met. So if the mapper doesn't add both tons of wandering NPC's plus lots of filtered collision objects that they'll practically run into, this should be reached very rarely.

Reply:
im really not comfortable with incremental performance losses in pathfinding.  we already have tickets out because it is slow enough.  if you are not going to implement the load-time checks for slugs, then drop this change and add a TODO

Reply:
I'd like to do this in map.py load if I could. The problem is that's called pretty early, there doesn't seem to be a way to fetch all NPC's present on the map at that time. Even if there was, what about NPC's that aren't added instantly but by an event after the map loaded?

I'll look a bit more if there's another way, but with the current implementation I just don't see one. I'll probably add that TODO instead.

Reply:
thats a good point, but i think that the databases are loaded before maps, in which case all the slugs should be available.  im trying to deprecate the ability to create custom npcs in the maps because we have some issues with it in the past, so we should rely on the slugs being available.

Reply:
Managed to fix this the right way! The validity of block parameters is now checked just once, each parameter being added to the list only if valid while an exception is raised otherwise. Let me know if this looks good in the latest commit.

Funny fact: My idea to check only the NPC's available on the map wouldn't have worked anyway. That's because an NPC might be added after the collision object is loaded. This would result in a crash as it won't find the NPC and there's no way of knowing what future events will do. So just making sure the NPC exists in the database was the right way to go all along.